### ISSUE TITLE
Start making release branches?

### ISSUE THREAD

**Issue by duk3luk3** [FAForever/db#118]
We have a lot of "feature" PRs pending for the db. Some PRs were merged and led to problems during the server upgrade last week, despite not having anything that depended on these PRs in a working state yet.

I think - especially with the inconsistent state of some of our database - we need a better process for this.

I propose:

1. Do not merge db PRs into master until API / Apps / Server are ready to use them
2. **Also** do not merge db PRs until migration has been tested using production data
3. **Also** do not merge db PRs until all components that interact with the db (java-api, python-api, server, ...?) have been tested using the result of the prod data migration

To make sure we do not lose too much velocity from this, I believe we should start making release branches:

1. Branch for production server (`release/prod`?), that has the db as it is running on prod
2. Branch for test server (`release/test`?), that has the db as it is "normally" running on the test server (for long-running tests, e.g. ice - although ice does not require db changes, this is just an example)
3. Branch for test server development (`release/develop`?) that can have essentially anything merged (cherry-picked) into it and can be deployed on the test server for testing new apps / new server versions

To make this not completely painful, we need some scripts / workflows that allow

* Grabbing a dump of prod db and restoring it into the test server db instance, with scripted censorship / swapping of credential data (e.g. user passwords, `oauth_clients` table)
* Applying the `release/test` migrations, with a quick way back to the old state (i.e. including a full dump) in case of errors
* Applying the `release/develop` migrations, also with a quick way back

This is **not** intended to make db migrations slow and painful. With the right scripting support, this will make db migrations and testing **less** painful. But most significantly, I believe this will allow us to be a lot safer and stop applying migrations blindly with uncertain results.

If people are in favour of this, I will start writing those scripts and working out process for this. I've already started some basic documentation: https://github.com/duk3luk3/faf-server-scripts/tree/master/testserver - I will be merging `faf-server-scripts` into `faftools` (which already has a scripts folder) and continue there.

**Comment by IDragonfire** [comment]
> 1. Do not merge db PRs into master until API / Apps / Server are ready to use them

Or backwards compatibility exists (e.g. a view)

> 2. Also do not merge db PRs until migration has been tested using production data

Should be solved with this PR https://github.com/FAForever/db/pull/112
Also see https://github.com/FAForever/db/issues/106

> 3. Also do not merge db PRs until all components that interact with the db (java-api, python-api, server, ...?) have been tested using the result of the prod data migration

Partial solved with https://github.com/FAForever/db/pull/112, but the main problem is, that nobody has the complete view of the all components. Maybe we should prepare a Pull Request Template

https://help.github.com/articles/creating-a-pull-request-template-for-your-repository/

# Template
### Please consider the affect on the following components
- [ ] python-server
- [ ] python-api
- [ ] java-api
- [ ] clanapp
- [ ] website

@duk3luk3 @micheljung 
Did I miss any direct consumer of the database?

I like the [release branching model](http://nvie.com/posts/a-successful-git-branching-model/), but for me view we have it already.

`develop` -> should be on the testserer
`master` -> next version on the prodserver
last tagged version -> currectn version on the prodserver

From my view the problem is, that we should migrate more or use travis for testing migrations^^

> To make this not completely painful, we need some scripts / workflows that allow 

Feel free to add scripts to this repo:
https://github.com/FAForever/faf-db-dump

PS.: Feel free to edit https://github.com/FAForever/db/wiki/How-to-Contribute

**Comment by duk3luk3** [comment]
Sadly we can't have "production-grade" data in the automatic tests, that is simply not feasible. This has to be done using manual testing.

Unless we start including test data - dumped and sanitized from prod db - with every PR and figure out a way to make the automatic build use them...

The faf-db-dump scripts are very good. I think they should also be merged into faftools.


>>  Do not merge db PRs into master until API / Apps / Server are ready to use them

> Or backwards compatibility exists (e.g. a view)

I think I may not have explained my reasoning here sufficiently. It's about simply not making changes to the production db that **are completely unecessary**, backwards compatibility or not. Every db change / migration has a possibility of failing or leaving something unable to use the db properly. So the migrations simply shouldn't be applied until we actually need them.

The correct time to apply a db change to prod is when you deploy a new app / new version to prod that needs that change.

**Comment by IDragonfire** [comment]
> The correct time to apply a db change to prod is when you deploy a new app / new version to prod that needs that change.

That is where we have different opinions, if we deploy as late as possible we have huge deployments ...
e.g. DB Update, API Update, App Update ... (see last update)

If one fail we cannot continue ... If we found an error later we must investigate all layers ...

It is better to deploy small updates frequently from my view.

Another problmem with your solution are long living branches across multiple repositories ...

**Comment by duk3luk3** [comment]
Hm, I see your point. And I agree, pushing db updates to prod without updating everything else at the same time would mean less possibility for failure.

**Comment by IDragonfire** [comment]
In terms of DevOps/Continous Delivery it is important to deploy often and fast,
this is the only way to train the developer and react on errors ...

The last update was a mess because we migrated from v22 to v34 ...
**12** Migrations  ...

**Comment by Brutus5000** [comment]
Downlord stops all accessing services before running the migration. I am not sure if this is technically enforced, but adding indices etc. doesn't seem to be something you would want to do on a system under load.
However stopping all servers means all current games get killed in terms of rating and replay. That is the reason, why we have the monthly release schedule.
Monthly should also be "deplyoing often" in terms of DevOps.

If the API gets more integrated into all clients one day we can't simply redeploy it as today without breaking client connections too. I don't think we can and should decouple this.

**Comment by micheljung** [comment]
I don't have time to read/respond to all of this, but I disagree with the very premise of this issue:
> Some PRs were merged and led to problems during the server upgrade last week, despite not having anything that depended on these PRs in a working state yet.

Hit me up for more details, or wait until I find time to explain

**Comment by micheljung** [comment]
> Do not merge db PRs into master until API / Apps / Server are ready to use them
In most cases, it's easy to prepare the database and build the application on top of it later.
But it's rarely easy to add functionality to the application without having the proper database, especially not when the feature depends on multiple unmerged PRs. And as dragonfire already said, the more changes go into one DB migration, the more time the migration takes and the higher the risk.

Therefore, I would continue "DB first".

> Also do not merge db PRs until migration has been tested using production data
While this would be desirable, I don't see it as viable. First of all, most people have no access to production data (and obviously shouldn't have) and even if we provide a sanitized dump, testing against production data is very time consuming as it takes about 20min just to import the current database, and another XX minutes to perform the migration (depending on it's nature). By the minute you start, your test is already obsolete as the production DB changed.

Therefore, I would instead continue with dragonfire's sanitized and reduced test dump.

> Also do not merge db PRs until all components that interact with the db (java-api, python-api, server, ...?) have been tested using the result of the prod data migration
This would be desirable as well but is a huge amount of work as well, especially if you want to test *all* relevant aspects (you might not even think of all of them, and some might not even be testable by yourself (e.g. ladder)). You'll never know if you tested everything.

So far, we looked at where the affected tables are used and decided, based on code analysis, if there's need for change or not. Problems we realized later on, like people with an expired ban not being able to log in, likely wouldn't have surfaced because nobody thought of logging in with a once banned user.

*If* we want more testing, then I strongly recommend building automated system tests with a deployed DB, server, api where real clients connects and simulate a users.


Next, the purpose of release branches (as of GitFlow) is to prepare a release, so they are short-lived and deleted as soon as a release has been built. I believe that the proposed Branching-Modell adds more confusion and chaos than it does help. It will be very hard to track which feature is on which branch and how to merge/cherry-pick between these branches.

* What's described as `release/prod` is exactly what `master` already is
* What's described as `release/test` pretty much reflect the purpose of `develop` - testing the next "development version". If we're testing a specific feature like ICE, then we'd rather just check out the respective feature-branch rather than an abstract `release/test` which is redundant with `develop`. Testing DB updates before performing it on prod, as we do it, reflects the purpose of a release branch as of GitFlow - so this is already covered as well.

>  I believe this will allow us to be a lot safer and stop applying migrations blindly with uncertain results
I've been performing DB updates for FAF in the past 1.5 years, we've already improved the process a lot and great care has always been taken. It's not making me happy to read that we're _applying migrations blindly_ with _uncertain results_ - based on some minor problems we had in the last one - from a person who hasn't done even a single update yet, especially not after all the efforts we've put into ensuring error-free migrations.
